File: studio/repositories/users_repo.py
"""User accounts and one-time auth tokens (verify / reset)."""

import hashlib
import secrets
import time

from sqlalchemy.exc import IntegrityError

from studio import db
from studio.models import AuthToken, User, to_doc

VERIFY_TTL_S = 24 * 3600
RESET_TTL_S = 3600


class AuthDbUnavailable(Exception):
    pass


class EmailTaken(Exception):
    pass


def _normalize_email(email):
    return (email or "").strip().lower()


def _token_hash(raw):
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()


def public(user):
    if not user:
        return user
    clean = dict(user)
    clean.pop("passwordHash", None)
    clean.pop("passwordSalt", None)
    return clean


def create(name, email, password_hash, salt):
    doc = User(
        userId=f"usr-{secrets.token_hex(8)}",
        email=_normalize_email(email),
        name=name.strip(),
        passwordHash=password_hash,
        passwordSalt=salt,
        emailVerified=False,
        createdAt=time.time(),
    )
    with db.session() as s:
        s.add(doc)
        try:
            s.commit()
        except IntegrityError as exc:
            s.rollback()
            raise EmailTaken(_normalize_email(email)) from exc
        return public(to_doc(doc))


def find_by_email(email):
    with db.session() as s:
        row = s.query(User).filter_by(email=_normalize_email(email)).first()
        return to_doc(row)


def find_by_id(user_id):
    with db.session() as s:
        row = s.query(User).filter_by(userId=user_id).first()
        return to_doc(row)


def is_empty():
    with db.session() as s:
        return s.query(User).first() is None


def set_verified(user_id):
    with db.session() as s:
        s.query(User).filter_by(userId=user_id).update({"emailVerified": True})
        s.commit()


def set_password(user_id, password_hash, salt):
    with db.session() as s:
        s.query(User).filter_by(userId=user_id).update(
            {"passwordHash": password_hash, "passwordSalt": salt}
        )
        s.commit()


def issue_token(user_id, purpose, ttl_s):
    raw = secrets.token_urlsafe(32)
    with db.session() as s:
        s.add(
            AuthToken(
                tokenHash=_token_hash(raw),
                userId=user_id,
                purpose=purpose,
                expiresAt=time.time() + ttl_s,
            )
        )
        s.commit()
    return raw


def consume_token(raw, purpose):
    hashed = _token_hash(raw or "")
    with db.session() as s:
        row = s.query(AuthToken).filter_by(tokenHash=hashed, purpose=purpose).first()
        if not row:
            return None
        expires_at, user_id = row.expiresAt, row.userId
        s.query(AuthToken).filter_by(tokenHash=hashed).delete()
        s.commit()
    return user_id if expires_at >= time.time() else None

File: src/server/repos/usersRepo.ts
import { Prisma } from "@/generated/prisma/client";
import { prisma } from "../db";
import { tokenHash } from "../securityNode";
import { toDoc } from "./doc";

export const VERIFY_TTL_S = 24 * 3600;
export const RESET_TTL_S = 3600;

export class AuthDbUnavailable extends Error {}
export class EmailTaken extends Error {}

export type UserDoc = Record<string, unknown> & {
  userId: string;
  email: string;
  name?: string | null;
  emailVerified?: boolean | null;
  passwordHash?: string | null;
  passwordSalt?: string | null;
};

function normalizeEmail(email: string): string {
  return (email ?? "").trim().toLowerCase();
}

export function publicUser(user: UserDoc | null): UserDoc | null {
  if (!user) return user;
  const clean = { ...user };
  delete clean.passwordHash;
  delete clean.passwordSalt;
  return clean;
}

export async function create(
  name: string,
  email: string,
  passwordHash: string,
  salt: string,
): Promise<UserDoc> {
  const userId = `usr-${randomHex(8)}`;
  try {
    const row = await prisma().user.create({
      data: {
        userId,
        email: normalizeEmail(email),
        name: name.trim(),
        passwordHash,
        passwordSalt: salt,
        emailVerified: false,
        createdAt: Date.now() / 1000,
      },
    });
    return publicUser(toDoc(row) as UserDoc)!;
  } catch (exc) {
    if (exc instanceof Prisma.PrismaClientKnownRequestError && exc.code === "P2002") {
      throw new EmailTaken(normalizeEmail(email));
    }
    throw exc;
  }
}

export async function findByEmail(email: string): Promise<UserDoc | null> {
  const row = await prisma().user.findUnique({ where: { email: normalizeEmail(email) } });
  return publicUser(toDoc(row) as UserDoc | null);
}

export async function findById(userId: string): Promise<UserDoc | null> {
  const row = await prisma().user.findUnique({ where: { userId } });
  return publicUser(toDoc(row) as UserDoc | null);
}

export async function getRole(userId: string): Promise<string | null> {
  const row = await prisma().user.findUnique({ where: { userId }, select: { role: true } });
  return row?.role ?? null;
}

export async function findByEmailWithSecrets(email: string): Promise<UserDoc | null> {
  const row = await prisma().user.findUnique({ where: { email: normalizeEmail(email) } });
  return toDoc(row) as UserDoc | null;
}

export async function isEmpty(): Promise<boolean> {
  const row = await prisma().user.findFirst({ select: { id: true } });
  return !row;
}

export async function setVerified(userId: string): Promise<void> {
  await prisma().user.updateMany({ where: { userId }, data: { emailVerified: true } });
}

export async function setPassword(
  userId: string,
  passwordHash: string,
  salt: string,
): Promise<void> {
  await prisma().user.updateMany({
    where: { userId },
    data: { passwordHash, passwordSalt: salt },
  });
}

export async function issueToken(userId: string, purpose: string, ttlS: number): Promise<string> {
  const raw = randomUrlSafe(32);
  await prisma().authToken.create({
    data: {
      tokenHash: tokenHash(raw),
      userId,
      purpose,
      expiresAt: Date.now() / 1000 + ttlS,
    },
  });
  return raw;
}

export async function consumeToken(raw: string, purpose: string): Promise<string | null> {
  const hashed = tokenHash(raw ?? "");
  const row = await prisma().authToken.findFirst({ where: { tokenHash: hashed, purpose } });
  if (!row) return null;
  const expiresAt = row.expiresAt ?? 0;
  const userId = row.userId ?? "";
  await prisma().authToken.deleteMany({ where: { tokenHash: hashed } });
  return expiresAt >= Date.now() / 1000 ? userId : null;
}

function randomHex(bytes: number): string {
  const buf = new Uint8Array(bytes);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}

function randomUrlSafe(bytes: number): string {
  const buf = new Uint8Array(bytes);
  crypto.getRandomValues(buf);
  return Buffer.from(buf).toString("base64url");
}
